File: app/src/main/java/com/ShaHar91/poems/ui/home/categories/CategoryViewModel.java
package com.shahar91.poems.ui.home.categories;

import com.shahar91.poems.data.models.Category;
import com.shahar91.poems.data.repositories.CategoryRepository;
import com.shahar91.poems.redux.AppState;
import com.shahar91.poems.redux.state.ViewState;
import com.shahar91.poems.ui.base.normal.BaseGoogleViewModel;
import com.shahar91.poems.ui.home.categories.redux.CategoryActions;
import com.shahar91.poems.ui.home.categories.redux.CategoryState;
import com.yheriatovych.reductor.Actions;
import com.yheriatovych.reductor.Store;
import com.yheriatovych.reductor.rxjava2.RxStore;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import be.appwise.core.extensions.logging.LoggingExtensionsKt;
import io.reactivex.Observable;
import kotlin.Unit;

public class CategoryViewModel extends BaseGoogleViewModel {

    @Inject
    CategoryViewModel(Store<AppState> store) {
        super(store);
    }

    public void getAllCategories() {
        CategoryRepository.getCategories(categories -> {
            CategoryActions categoryActions = Actions.from(CategoryActions.class);
            store.dispatch(categoryActions.setCategoryList(new ArrayList<>(categories)));

            return Unit.INSTANCE;
        }, throwable -> {
            throwable.printStackTrace();
            LoggingExtensionsKt.loge(null, throwable, "");
            return Unit.INSTANCE;
        });
    }

    public Observable<List<Category>> categoriesStateListener() {
        Observable<AppState> store = RxStore.asObservable(this.store);
        return store.map(AppState::viewState)
                .map(ViewState::categoryState)
                .filter(categoryState -> categoryState.categoryList() != null)
                .map(CategoryState::categoryList)
                .distinctUntilChanged();
    }
}

File: app/src/main/java/com/ShaHar91/poems/ui/home/categories/CategoryViewModel.kt
package com.shahar91.poems.ui.home.categories

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.shahar91.poems.domain.repository.ICategoryRepository
import com.shahar91.poems.ui.base.PoemBaseViewModel

class CategoryViewModel(
    private val categoryRepository: ICategoryRepository
) : PoemBaseViewModel() {

    private val _isRefreshing = MutableLiveData(false)
    val isRefreshing: LiveData<Boolean> get() = _isRefreshing
    fun setIsRefreshing(refreshing: Boolean) {
        _isRefreshing.postValue(refreshing)
    }

    var categoriesLive = categoryRepository.findAllLive()

    init {
        getAllCategories()
    }

    fun getAllCategories() = launchAndLoad {
        categoryRepository.fetchCategories()
        setIsRefreshing(false)
    }
}